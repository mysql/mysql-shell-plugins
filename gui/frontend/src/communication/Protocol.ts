/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

/* eslint-disable max-len */
/* eslint-disable @typescript-eslint/naming-convention */

import { ListenerEntry } from "../supplement/Dispatch";
import { uuid } from "../utilities/helpers";

export type ShellDictionaryType = string | number | boolean | undefined | unknown | null | IShellDictionary;
export interface IShellDictionary {
    [key: string]: ShellDictionaryType | ShellDictionaryType[];
}

export enum ShellPromptResponseType {
    Ok = "OK",
    Cancel = "CANCEL"
}

// The request structure sent to the backend. That's why it uses snake case.
export interface IShellRequest extends IShellDictionary {
    request_id: string; // A unique ID to identify this request. It's used for all responses.
    request: string;    // The requested operation.

    command?: string;        // Optional field to carry the command if this is an execution request.
    args?: IShellDictionary; // Optional arguments for the command.
}

//  Begin auto generated types

//  End auto generated types

export class Protocol {

    public static getRequestCommandExecute(command: string, rest: IShellDictionary): IShellRequest {
        return Protocol.getStandardRequest("execute", { command, ...rest });
    }

    public static getRequestUsersAuthenticate(username: string, password: string): IShellRequest {
        return Protocol.getStandardRequest("authenticate", {
            username,
            password,
        });
    }

    public static getRequestPromptReply(request_id: string, type: ShellPromptResponseType, reply: string, moduleSessionId: string): IShellRequest {
        const result = Protocol.getStandardRequest("prompt_reply", {
            request_id,
            type,
            reply,
            module_session_id: moduleSessionId,
        });

        return result;
    }

    public static getStandardRequest(request: string, rest?: IShellDictionary): IShellRequest {
        return {
            request_id: uuid(),
            request,
            ...rest,
        };
    }
}

export interface IPromptReplyBackend {
    sendReply: (requestId: string, type: ShellPromptResponseType, reply: string) => ListenerEntry;
}
