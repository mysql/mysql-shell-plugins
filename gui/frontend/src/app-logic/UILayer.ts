/*
 * Copyright (c) 2024, 2025, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import type { IStatusBarItem, StatusBarAlignment } from "../components/ui/Statusbar/StatusBarItem.js";
import type { IServicePasswordRequest } from "./general-types.js";

/**
 * The UI layer object comprises functionality that is specific to the environment it is used in.
 * This is only relevant for code which is shared between different platforms (e.g. the extension and the app).
 * The UI layer redirects requests for UI specific functionality to the appropriate implementation and must be set
 * up by the application logic layer (either the app host or the extension host).
 */
export interface IUILayer {
    /**
     * Shows an information notification. The message is automatically hidden after the given timeout
     * (in milliseconds, default is 5000).
     */
    showInformationNotification(message: string, timeout?: number): Promise<string | undefined>;

    /** Shows a warning notification. */
    showWarningNotification(message: string): Promise<string | undefined>;

    /** Shows a error notification. */
    showErrorNotification(message: string): Promise<string | undefined>;

    /**
     * Creates a status bar {@link StatusBarItem item}.
     *
     * @param alignment The alignment of the item.
     * @param priority The priority of the item. Higher values mean the item should be shown more to the left.
     * @returns A new status bar item.
     */
    createStatusBarItem(alignment?: StatusBarAlignment, priority?: number): IStatusBarItem;

    /**
     * Creates a status bar {@link StatusBarItem item}.
     *
     * @param id The unique identifier of the item.
     * @param alignment The alignment of the item.
     * @param priority The priority of the item. Higher values mean the item should be shown more to the left.
     * @returns A new status bar item.
     */
    createStatusBarItem(id: string, alignment?: StatusBarAlignment, priority?: number): IStatusBarItem;

    /**
     * Shows a message in the status bar.
     *
     * @param message The message to show.
     * @param timeout The time in milliseconds after which the message is automatically hidden. If not set, the message
     *               is shown until the next message is set.
     */
    setStatusBarMessage(message: string, timeout?: number): void;

    /**
     * Asks the user for a yes/no decision. Returns the yes/no string, depending on what the user clicked or
     * undefined if the user clicked outside/cancelled the dialog (e.g. pressed <escape>).
     */
    confirm(message: string, yes: string, no: string): Promise<string | undefined>;

    /**
     * Triggers a password request.
     *
     * @param request The request with details about the password prompt.
     *
     * @returns The entered password or undefined if the password dialog was cancelled.
     */
    requestPassword(request: IServicePasswordRequest): Promise<string | undefined>;
}

const defaultHandler = (): number => {
    console.error("Internal error: the UI layer has not been set up.");

    return -1;
};

const asyncDefaultHandler = (): Promise<string | undefined> => {
    console.error("Internal error: the UI layer has not been set up.");

    return Promise.resolve(undefined);
};

export let ui: IUILayer = {
    showInformationNotification: asyncDefaultHandler,
    showWarningNotification: asyncDefaultHandler,
    showErrorNotification: asyncDefaultHandler,
    createStatusBarItem: (): IStatusBarItem => {
        console.error("Internal error: the UI layer has not been set up.");

        return {} as IStatusBarItem;
    },
    setStatusBarMessage: defaultHandler,
    confirm: asyncDefaultHandler,
    requestPassword: asyncDefaultHandler,
};

export const registerUiLayer = (newLayer: IUILayer): void => {
    ui = newLayer;
};
