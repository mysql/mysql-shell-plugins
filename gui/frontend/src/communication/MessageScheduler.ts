/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { existsSync, readFileSync } from "fs";
import { join } from "path";
import { default as NodeWebSocket } from "ws";

import { CommunicationEvents, IShellRequest, IWebSessionData } from ".";
import { appParameters, requisitions } from "../supplement/Requisitions";
import {
    dispatcher, IDispatchEventContext, ListenerEntry, eventFilterNoRequests,
} from "../supplement/Dispatch";
import { convertSnakeToCamelCase } from "../utilities/helpers";

export enum ConnectionEventType {
    Open = 1,
    Close,
    Error,
    Message,
}

/** Wrapper around a web socket that performs (re)connection and message scheduling. */
export class MessageScheduler {
    private static instance?: MessageScheduler;

    private debugging = false;

    private connectionEstablished = false;
    private autoReconnecting = false;
    private disconnecting = false;

    private reconnectTimer: ReturnType<typeof setTimeout> | null;

    private socket?: WebSocket | NodeWebSocket;

    public static get get(): MessageScheduler {
        if (!MessageScheduler.instance) {
            MessageScheduler.instance = new MessageScheduler();
        }

        return MessageScheduler.instance;
    }

    private constructor() {
        // Singleton pattern.
    }

    /**
     * @returns True if the connection is established.
     */
    public get isConnected(): boolean {
        if (this.socket) {
            return this.connectionEstablished && this.socket.readyState === this.socket.OPEN;
        }

        return false;
    }

    public set inDebugCall(value: boolean) {
        this.debugging = value;
    }

    /**
     * Opens a web socket connection.
     *
     * @param url The target to connect to. This is a normal http(s) target from which the websocket address is
     *            derived.
     * @param shellConfigDir The current shell configuration folder, which can be different, depending on the
     *                       current host. Only needed when running from the extension (under Node.js).
     *
     * @returns A promise which indicates success or failure.
     */
    public connect(url: URL, shellConfigDir: string): Promise<void> {
        this.disconnecting = false;

        return new Promise((resolve, reject) => {
            // If already open or connecting don't do anything.
            if (this.socket && (this.socket.readyState === this.socket.OPEN ||
                this.socket.readyState === this.socket.CONNECTING)) {
                resolve();
            } else {
                url.protocol = url.protocol.replace("http", "ws"); // ws or wss
                url.pathname = "ws1.ws";

                /* istanbul ignore next */
                if (appParameters.inDevelopment) {
                    url.port = "8000";
                }

                if (typeof WebSocket !== "undefined") {
                    const socket = new WebSocket(url.toString());

                    socket.addEventListener("close", this.onClose);
                    socket.addEventListener("message", this.onMessage);
                    socket.addEventListener("open", this.onOpen.bind(this, resolve));
                    socket.addEventListener("error", this.onError.bind(this, reject));

                    this.socket = socket;
                } else {
                    const caFile = join(shellConfigDir, "plugin_data/gui_plugin/web_certs/rootCA.crt");

                    let ca;
                    if (existsSync(caFile)) {
                        ca = readFileSync(caFile);
                    }

                    const socket = new NodeWebSocket(url.toString(), { ca });

                    socket.addEventListener("close", this.onClose);
                    socket.addEventListener("message", this.onMessage);
                    socket.addEventListener("open", this.onOpen.bind(this, resolve));
                    socket.addEventListener("error", this.onError.bind(this, reject));

                    this.socket = socket;
                }
            }
        });
    }

    /**
     * Disconnect the socket
     */
    public disconnect(): void {
        if (this.socket) {
            try {
                this.disconnecting = true;
                this.socket.close(); // Careful when specifying a code here. It must be valid or the socket will hang.
                delete this.socket;
            } catch (e) {
                /* istanbul ignore next */
                console.error("Internal error while closing websocket: " + String(e));
            }
        }
    }

    /**
     * Send a request to the shell. Create a listener so that the caller can wait for the response.
     *
     * @param request The request to send to the shell. It must use snake case for its member fields.
     * @param context Context data to supply to the listener when the response arrives.
     * @returns a listener that waits for the shell response
     */
    public sendRequest(request: IShellRequest, context: IDispatchEventContext): ListenerEntry {
        dispatcher.mapMessageContext(request.request_id, context);
        const listener = ListenerEntry.createByID(request.request_id, { filters: [eventFilterNoRequests] });

        this.socket?.send(JSON.stringify(request));

        if (!this.debugging) {
            dispatcher.triggerEvent(CommunicationEvents.generateRequestEvent(request));
        }

        return listener;
    }

    public clearState(): void {
        // no-op
    }

    /**
     * Called when the socket is opened.
     *
     * @param resolve The resolver function to signal when the socket connection is available.
     */
    private onOpen = (resolve: (value: void | PromiseLike<void>) => void): void => {
        if (this.autoReconnecting) {
            this.autoReconnecting = false;

            // TODO: convert this error message to a message toast.
            void requisitions.execute("showError", [
                "Connection Recovering",
                "The connection was automatically re-establish after a failure.",
            ]);
        }

        if (this.reconnectTimer) {
            // Clear the timer for the disconnection message, since the connection is back.
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = null;
        }

        this.connectionEstablished = true;
        dispatcher.triggerNotification("socketOpened");
        resolve();
    };


    /**
     * Called when the socket is closed or failed to open. Unfortunately, the given close code is always the same
     * (1006) and the close reason is always empty.
     */
    private onClose = (): void => {
        const wasConnected = this.connectionEstablished;

        this.connectionEstablished = false;
        this.autoReconnecting = false;
        dispatcher.triggerNotification("socketClosed");

        if (!this.disconnecting && !this.debugging) {
            if (wasConnected) {
                if (this.reconnectTimer) {
                    clearTimeout(this.reconnectTimer);
                }

                // Show the disconnection error only after 3 seconds, to allow automatic reconnection without disturbing
                // the user (which can also happen when the computer went to sleep).
                this.reconnectTimer = setTimeout(() => {
                    this.reconnectTimer = null;
                    void requisitions.execute("showError", [
                        "Browser Connection Error",
                        "The connection was interrupted unexpectedly. An automatic reconnection failed, but the " +
                        "application will continue trying to reconnect.",
                    ]);
                }, 3000);
            }
        }
    };

    private onMessage = (event: MessageEvent | NodeWebSocket.MessageEvent): void => {
        // Convert message data string to object and also convert from snake case to camel case.
        // However, ignore row data in this process, as it either comes as array (no keys to convert) or comes
        // as object, whose keys are the real column names (which must stay as is).
        if (event.data) { // Sanity check. Should never happen.
            const data = JSON.parse(event.data as string) as object;
            const serverData = convertSnakeToCamelCase(data, { ignore: ["rows"] }) as IWebSessionData;

            // Note: the context given here by the event factories is replaced by the one specified for a specific
            //       request in the sendRequest call. Only in error cases the context from here might be used.
            if (serverData.sessionUuid) {
                dispatcher.triggerEvent(CommunicationEvents.generateWebSessionEvent(serverData), this.debugging);
            } else {
                dispatcher.triggerEvent(CommunicationEvents.generateResponseEvent("serverResponse", serverData),
                    this.debugging);
            }
        }
    };

    /**
     * Called when an error occurred in the socket connection.
     *
     * @param reject The reject function to call to signal that the connection had a failure to open.
     * @param event Event that gives more info about the error.
     */
    private onError = (reject: (reason?: unknown) => void, event: Event | NodeWebSocket.ErrorEvent): void => {
        if ((event as NodeWebSocket.ErrorEvent).message) {
            reject((event as NodeWebSocket.ErrorEvent).message);
        } else {
            reject(String(event));
        }

        void requisitions.execute("showError", [
            "Communication Error",
            "Could not establish a connection to the backend.",
        ]);
    };
}
