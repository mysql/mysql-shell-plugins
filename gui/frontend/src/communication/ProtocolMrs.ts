/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

export enum ShellAPIMrs {
    //  Begin auto generated API names
    MrsAddService = "mrs.add.service",
    MrsGetService = "mrs.get.service",
    MrsListServices = "mrs.list.services",
    MrsEnableService = "mrs.enable.service",
    MrsDisableService = "mrs.disable.service",
    MrsDeleteService = "mrs.delete.service",
    MrsSetServiceDefault = "mrs.set.service.default",
    MrsSetServiceContextPath = "mrs.set.service.context_path",
    MrsSetServiceProtocol = "mrs.set.service.protocol",
    MrsSetServiceComments = "mrs.set.service.comments",
    MrsSetServiceOptions = "mrs.set.service.options",
    MrsUpdateService = "mrs.update.service",
    MrsGetServiceRequestPathAvailability = "mrs.get.service_request_path_availability",
    MrsAddSchema = "mrs.add.schema",
    MrsGetSchema = "mrs.get.schema",
    MrsListSchemas = "mrs.list.schemas",
    MrsEnableSchema = "mrs.enable.schema",
    MrsDisableSchema = "mrs.disable.schema",
    MrsDeleteSchema = "mrs.delete.schema",
    MrsSetSchemaName = "mrs.set.schema.name",
    MrsSetSchemaRequestPath = "mrs.set.schema.request_path",
    MrsSetSchemaRequiresAuth = "mrs.set.schema.requires_auth",
    MrsSetSchemaItemsPerPage = "mrs.set.schema.items_per_page",
    MrsSetSchemaComments = "mrs.set.schema.comments",
    MrsUpdateSchema = "mrs.update.schema",
    MrsAddContentSet = "mrs.add.content_set",
    MrsListContentSets = "mrs.list.content_sets",
    MrsGetContentSet = "mrs.get.content_set",
    MrsEnableContentSet = "mrs.enable.content_set",
    MrsDisableContentSet = "mrs.disable.content_set",
    MrsDeleteContentSet = "mrs.delete.content_set",
    MrsAddDbObject = "mrs.add.db_object",
    MrsGetDbObject = "mrs.get.db_object",
    MrsGetDbObjectRowOwnershipFields = "mrs.get.db_object_row_ownership_fields",
    MrsGetDbObjectFields = "mrs.get.db_object_fields",
    MrsListDbObjects = "mrs.list.db_objects",
    MrsGetDbObjectParameters = "mrs.get.db_object_parameters",
    MrsSetDbObjectRequestPath = "mrs.set.db_object.request_path",
    MrsSetDbObjectCrudOperations = "mrs.set.db_object.crud_operations",
    MrsEnableDbObject = "mrs.enable.db_object",
    MrsDisableDbObject = "mrs.disable.db_object",
    MrsDeleteDbObject = "mrs.delete.db_object",
    MrsUpdateDbObject = "mrs.update.db_object",
    MrsListContentFiles = "mrs.list.content_files",
    MrsGetAuthenticationVendors = "mrs.get.authentication_vendors",
    MrsAddAuthenticationApp = "mrs.add.authentication_app",
    MrsListAuthenticationApps = "mrs.list.authentication_apps",
    MrsInfo = "mrs.info",
    MrsVersion = "mrs.version",
    MrsLs = "mrs.ls",
    MrsConfigure = "mrs.configure",
    MrsStatus = "mrs.status",
    //  End auto generated API names
}
