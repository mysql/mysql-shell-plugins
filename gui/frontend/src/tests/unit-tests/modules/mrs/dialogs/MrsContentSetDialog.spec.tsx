/*
 * Copyright (c) 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

/* eslint-disable dot-notation */

import { createRef } from "preact";

import { mount } from "enzyme";

import { registerUiLayer } from "../../../../../app-logic/UILayer.js";
import { DataCallback } from "../../../../../communication/MessageScheduler.js";
import { IShellDictionary } from "../../../../../communication/Protocol.js";
import { IMrsAddContentSetData, ShellAPIMrs } from "../../../../../communication/ProtocolMrs.js";
import { MrsHub } from "../../../../../modules/mrs/MrsHub.js";
import { ShellInterfaceSqlEditor } from "../../../../../supplement/ShellInterface/ShellInterfaceSqlEditor.js";
import { MySQLShellLauncher } from "../../../../../utilities/MySQLShellLauncher.js";
import { KeyboardKeys } from "../../../../../utilities/helpers.js";
import { uiLayerMock } from "../../../__mocks__/UILayerMock.js";
import {
    DialogHelper, JestReactWrapper, createBackend, recreateMrsData, sendKeyPress, setupShellForTests,
} from "../../../test-helpers.js";

describe("MRS Content Set dialog tests", () => {
    let host: JestReactWrapper;
    // let service: IMrsServiceData;
    let launcher: MySQLShellLauncher;
    const hubRef = createRef<MrsHub>();
    let dialogHelper: DialogHelper;
    let backend: ShellInterfaceSqlEditor;

    beforeAll(async () => {
        registerUiLayer(uiLayerMock);
        launcher = await setupShellForTests(false, true, "DEBUG2");

        await recreateMrsData();

        host = mount<MrsHub>(<MrsHub ref={hubRef} />);

        dialogHelper = new DialogHelper("mrsContentSetDialog", "MRS Content Set");
    });

    afterAll(async () => {
        await backend.execute("DROP DATABASE IF EXISTS mysql_rest_service_metadata");
        await backend.execute("DROP DATABASE IF EXISTS MRS_TEST");
        await backend.closeSession();
        await launcher.exitProcess();
        host.unmount();
    });

    beforeEach(async () => {
        backend = await createBackend();
    });

    it("Show MRS Content Set Dialog (snapshot) and escape", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);

        const promise = hubRef.current!.showMrsContentSetDialog(backend);
        await dialogHelper.waitForDialog();

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(1);

        expect(portals[0]).toMatchSnapshot();

        setTimeout(() => {
            sendKeyPress(KeyboardKeys.Escape);
        }, 250);

        await promise;

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
    });

    it("Show MRS Content Set Dialog and cancel", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);

        const promise = hubRef.current!.showMrsContentSetDialog(backend);
        await dialogHelper.waitForDialog();

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(1);

        await dialogHelper.clickCancel();

        await promise;

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
    });

    it("Dialog error testing", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);

        backend.mrs.addContentSet = (contentDir: string, requestPath: string,
            requiresAuth: boolean, options: IShellDictionary | null,
            serviceId?: string, comments?: string,
            enabled?: number, replaceExisting?: boolean,
            ignoreList?: string,
            callback?: DataCallback<ShellAPIMrs.MrsAddContentSet>): Promise<IMrsAddContentSetData> => {
            expect(requestPath).toBe("/someRequestPath");

            expect(contentDir).toBeDefined();
            expect(requiresAuth).toBeDefined();
            expect(options).toBeDefined();
            expect(serviceId).toBeDefined();
            expect(comments).toBeDefined();
            expect(enabled).toBeDefined();
            expect(replaceExisting).toBeDefined();
            expect(callback).toBeDefined();

            return Promise.resolve({});
        };

        const promise = hubRef.current!.showMrsContentSetDialog(backend);
        await dialogHelper.waitForDialog();

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(1);

        await dialogHelper.setInputText("requestPath", "");
        await dialogHelper.clickOk();
        dialogHelper.verifyErrors(["The request path must not be empty."]);

        await dialogHelper.setInputText("requestPath", "someRequestPath");
        await dialogHelper.clickOk();
        dialogHelper.verifyErrors(["The request path must start with /."]);

        await dialogHelper.setInputText("requestPath", "/someRequestPath");
        await dialogHelper.clickOk();
        dialogHelper.verifyErrors();

        await promise;

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
    });
});
