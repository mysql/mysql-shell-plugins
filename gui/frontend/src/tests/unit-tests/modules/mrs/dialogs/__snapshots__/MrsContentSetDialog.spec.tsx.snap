// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MRS Content Set dialog tests Show MRS Content Set Dialog (snapshot) and escape 1`] = `
<div
  class="msg portal"
  id="mrsContentSetDialog"
  style="--background-opacity: 0.5;"
>
  <div
    class="msg dialog visible valueEditDialog"
  >
    <div
      class="title verticalCenterContent"
    >
      <div
        class="msg icon codicon codicon-terminal"
      />
      <label
        class="msg label"
        data-tooltip="expand"
      >
        MRS Content Set
      </label>
      <div
        class="msg button imageOnly"
        id="closeButton"
        role="button"
        tabindex="0"
      >
        <div
          class="msg icon"
          style="mask-image: url("close2.svg");"
        />
      </div>
    </div>
    <div
      class="content fixedScrollbar"
    >
      <div
        class="msg grid"
        id="contentGrid"
        style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
      >
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: row; grid-column: span 8;"
        >
          <label
            class="msg label sectionTitle"
            data-tooltip="expand"
          >
            Enter Configuration Values for the New MRS Static Content Set
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 3;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Request Path
          </label>
          <div
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              class="msg input valueEditor"
              id="requestPath"
              spellcheck="true"
              type="text"
            />
          </div>
          <label
            class="msg label cellDescription"
            data-tooltip="expand"
          >
            The request path to access the content, has to start with /
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 3;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            REST Service Path
          </label>
          <div
            class="msg container fixedScrollbar dropdown valueEditor"
            data-tooltip="inherit"
            id="service"
            style="flex-direction: row; align-items: center;"
            tabindex="0"
          >
            <label
              class="msg label ellipsis"
              data-tooltip="expand"
            >
              /myService
            </label>
          </div>
          <label
            class="msg label cellDescription"
            data-tooltip="expand"
          >
            The MRS Service to hold the content
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 2;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Access Control
          </label>
          <div
            class="msg container fixedScrollbar dropdown valueEditor"
            data-tooltip="inherit"
            id="enabled"
            style="flex-direction: row; align-items: center;"
            tabindex="0"
          >
            <label
              class="msg label ellipsis"
              data-tooltip="expand"
            >
              Access ENABLED
            </label>
          </div>
          <label
            class="msg checkbox unchecked valueEditor"
            id="requiresAuth"
            tabindex="0"
          >
            <span
              class="checkMark"
            />
            Auth. Required
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 5;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Folder to upload
          </label>
          <div
            class="msg container fixedScrollbar fileSelector valueEditor"
            style="flex-direction: row;"
          >
            <input
              class="msg input"
              id="directory"
              spellcheck="true"
              type="text"
            />
          </div>
          <label
            class="msg label cellDescription"
            data-tooltip="expand"
          >
            The folder that should be uploaded, including all its files and sub-folders
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 3;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Files to ignore
          </label>
          <div
            class="msg container fixedScrollbar inputWithProgress"
            style="flex-direction: row;"
          >
            <input
              class="msg input valueEditor"
              id="ignoreList"
              spellcheck="true"
              type="text"
            />
          </div>
          <label
            class="msg label cellDescription"
            data-tooltip="expand"
          >
            A list of files to ignore, use * and ? as wildcards
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: row; grid-column: span 8;"
        >
          <div
            class="msg container fixedScrollbar tabview top"
            style="flex-direction: column; --tabItem-border-width: 0px; --content-separator-width: 0px;"
          >
            <div
              class="msg container tabAreaContainer"
              style="flex-direction: row;"
            >
              <div
                class="scrollable"
              >
                <div
                  class="msg container tabArea stretched"
                  style="flex-direction: row;"
                >
                  <div
                    class="msg button tabItem selected"
                    id="page0"
                    role="button"
                    tabindex="-1"
                  >
                    <label
                      class="msg label"
                      data-tooltip="inherit"
                    >
                      Settings
                    </label>
                  </div>
                  <div
                    class="msg button tabItem"
                    id="page1"
                    role="button"
                    tabindex="-1"
                  >
                    <label
                      class="msg label"
                      data-tooltip="inherit"
                    >
                      Options
                    </label>
                  </div>
                </div>
                <div
                  class="scrollbar"
                >
                  <div
                    class="slider"
                  />
                </div>
              </div>
            </div>
            <div
              class="msg container fixedScrollbar tabContent"
              style="flex-direction: column;"
            >
              <div
                class="msg grid"
                style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
              >
                <div
                  class="msg container fixedScrollbar gridCell"
                  style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
                >
                  <label
                    class="msg label valueTitle"
                    data-tooltip="expand"
                  >
                    Comments
                  </label>
                  <div
                    class="msg container fixedScrollbar inputWithProgress"
                    style="flex-direction: row;"
                  >
                    <textarea
                      class="msg input valueEditor"
                      id="comments"
                      rows="3"
                      spellcheck="true"
                      type="text"
                    />
                  </div>
                </div>
                <div
                  class="msg container fixedScrollbar gridCell"
                  style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 4;"
                >
                  <label
                    class="msg label valueTitle"
                    data-tooltip="expand"
                  >
                    MRS Scripts
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="footer verticalCenterContent"
    >
      <div
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <div
          caption="OK"
          class="msg button"
          id="ok"
          role="button"
          tabindex="0"
        >
          OK
        </div>
        <div
          caption="Cancel"
          class="msg button"
          id="cancel"
          role="button"
          tabindex="0"
        >
          Cancel
        </div>
      </div>
    </div>
  </div>
</div>
`;
